### PR TITLE
Release Candidate azblob 0.4.0

### PR DESCRIPTION
**Pull request by mohsha-msft**
- [ ] The purpose of this PR is explained in this or a referenced issue.
- [ ] The PR does not update generated files.
   - These files are managed by the codegen framework at [Azure/autorest.go][].
- [ ] Tests are included and/or updated for code changes.
- [ ] Updates to [CHANGELOG.md][] are included.
- [ ] MIT license headers are included in each file.

[Azure/autorest.go]: https://github.com/Azure/autorest.go
[CHANGELOG.md]: https://github.com/Azure/azure-sdk-for-go/blob/main/CHANGELOG.md

### CODE REVIEW

File: sdk/storage/azblob/testdata/perf/go.mod [hunk 1]
```diff
@@ -1,17 +1,17 @@
 module github.com/Azure/azure-sdk-for-go/sdk/storage/azblob/testdata/perf
 
-go 1.17
+go 1.18
 
 replace github.com/Azure/azure-sdk-for-go/sdk/internal => ../../../../internal
```
Comment (jhendrixMSFT) [inline sdk/storage/azblob/testdata/perf/go.mod:5]: Can this be removed?

File: sdk/storage/azblob/README.md [hunk 2]
```diff
@@ -205,91 +205,91 @@ Three different clients are provided to interact with the various components of
 ### Example
 
 ```go
-	// Use your storage account's name and key to create a credential object, used to access your account.
-	// You can obtain these details from the Azure Portal.
```
[the remark this answers is not in the document]
Comment (jhendrixMSFT) [reply sdk/storage/azblob/README.md:209]: Fixed.

File: sdk/storage/azblob/testdata/perf/go.mod (hunk 1 above)
Comment (jhendrixMSFT) [reply sdk/storage/azblob/testdata/perf/go.mod]: Fixed.

File: sdk/storage/azblob/README.md [hunk 3]
```diff
@@ -302,10 +304,8 @@ addition, you can investigate the raw response of any response object:
 ```golang
 var errResp azcore.HTTPResponse
 resp, err := serviceClient.CreateContainer(context.Background(), "testcontainername", nil)
-if err != nil {
-   if errors.As(err, &errResp) {
-        // do something with errResp.RawResponse()
-   }
+if err != nil && errors.As(err, &errResp) {
+    // do something with errResp.RawResponse()
```
Comment (seankane-msft) [inline sdk/storage/azblob/README.md:308]: @jhendrixMSFT does `var errResp azcore.HTTPResponse` have to be a pointer? For example, in tests: https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/data/aztables/client_test.go#L30-L31. I suspect this shows up in other places.

File: sdk/storage/azblob/constants.go [hunk 4]
```diff
@@ -6,33 +6,49 @@
 
 package azblob
 
+var SASVersion = "2019-12-12"
```
Comment (seankane-msft) [inline sdk/storage/azblob/constants.go:9]: should this be a constant?

File: sdk/storage/azblob/constants.go [hunk 5]
```diff
@@ -6,33 +6,49 @@
...
+	headerContentEncoding   = "Content-Encoding"
+	headerContentLanguage   = "Content-Language"
+	headerContentType       = "Content-Type"
+	headerContentMD5        = "Content-MD5"
+	headerIfModifiedSince   = "If-Modified-Since"
+	headerIfMatch           = "If-Match"
+	headerIfNoneMatch       = "If-None-Match"
+	headerIfUnmodifiedSince = "If-Unmodified-Since"
+	headerRange             = "Range"
+	//headerDate                    = "Date"
```
Comment (seankane-msft) [inline sdk/storage/azblob/constants.go:39]: nit, but we can remove this commented out code

File: sdk/storage/azblob/doc.go [hunk 6]
```diff
@@ -105,13 +105,13 @@ There are three different clients provided to interact with the various componen
     * Get and set account settings.
     * Query, create, and delete containers within the account.
 
-2. **`ContainerClient`**
+2. **`containerClient`**
```
Comment (seankane-msft) [inline sdk/storage/azblob/doc.go:108]: why did this and appendBlobClient get lowercased?

File: sdk/storage/azblob/README.md (hunk 3 above)
Comment (jhendrixMSFT) [reply sdk/storage/azblob/README.md:308]: It depends.  If the error is returned by pointer, then you need to provide the error var by reference (i.e. pointer-to-pointer) in the call to `errors.As()`.

The sample is out-of-date though as `azcore.HTTPResponse` has been removed.  I'll get this fixed.

File: sdk/storage/azblob/constants.go (hunk 4 above)
Comment (jhendrixMSFT) [reply sdk/storage/azblob/constants.go:9]: I would assume so given that this is defined in `constants.go`.  Unless it's intended to be changed by the caller.  @mohsha-msft can you comment?

File: sdk/storage/azblob/highlevel.go [hunk 7]
```diff
@@ -59,17 +59,17 @@ type HighLevelUploadToBlockBlobOption struct {
 	TransactionalContentMD5 *[]byte
 }
 
-func (o *HighLevelUploadToBlockBlobOption) getStageBlockOptions() *StageBlockOptions {
-	leaseAccessConditions, _ := o.BlobAccessConditions.pointers()
-	return &StageBlockOptions{
+func (o *HighLevelUploadToBlockBlobOption) getStageBlockOptions() *BlockBlobStageBlockOptions {
```
Comment (seankane-msft) [inline sdk/storage/azblob/highlevel.go:62]: If this is a renaming (`StageBlockOptions` to `BlockBlobStageBlockOptions`), this should be in the changelog.

File: sdk/storage/azblob/zc_blob_client.go [hunk 8]
```diff
@@ -139,91 +161,91 @@ func (b *BlobClient) Undelete(ctx context.Context) (BlobUndeleteResponse, error)
...
 
-	return resp, handleError(err)
+	return toBlobSetTierResponse(resp), handleError(err)
 }
 
 // GetProperties returns the blob's properties.
 // For more information, see https://docs.microsoft.com/rest/api/storageservices/get-blob-properties.
-func (b *BlobClient) GetProperties(ctx context.Context, options *GetBlobPropertiesOptions) (GetBlobPropertiesResponse, error) {
-	basics, lease, cpk, access := options.pointers()
+func (b *BlobClient) GetProperties(ctx context.Context, options *BlobGetPropertiesOptions) (BlobGetPropertiesResponse, error) {
```
Comment (seankane-msft) [inline sdk/storage/azblob/zc_blob_client.go:173]: Same comment here about breaking changes. Changing the type of options is a breaking change and should be documented in CHANGELOG.md.

We had a ton of changes in other libraries and we're going to miss a few, but only showing one breaking change here is a bad look.

File: sdk/storage/azblob/constants.go (hunk 4 above)
Comment (mohsha-msft) [reply sdk/storage/azblob/constants.go:9]: Yes, having the SASVersion as var can give caller the option to change it if needed without making change in the code.

File: sdk/storage/azblob/constants.go (hunk 5 above)
Comment (mohsha-msft) [reply sdk/storage/azblob/constants.go:39]: Will be required in the code that I'm planning to add. So I wish to keep it here.

File: sdk/storage/azblob/highlevel.go (hunk 7 above)
Comment (mohsha-msft) [reply sdk/storage/azblob/highlevel.go]: There are a lot of such changes. It'll not be feasible to list them all. So, I've added one liner in changelog to address this.

**Comment by mohsha-msft** [comment]
This PR was still work in progress and I should have made it a draft instead.  

I've merged [PR 17647](https://github.com/Azure/azure-sdk-for-go/pull/17647) in this PR to add changelogs, write more examples, and some code reformatting. It should address the above-mentioned concerns. 

Apart from UDK changes, we are ready with the next preview release. Given that we're close to code-completion date, I'll try to add this in next release.

Please let me know if it is still missing anything.

**Review by jhendrixMSFT — APPROVED** [review]

**Pull request merged**